app/src/org/commcare/android/nfc/NdefRecordUtil.java

```diff
@@ -0,0 +1,133 @@
+            case NdefRecord.TNF_EXTERNAL_TYPE:
+                return handleExternalTypeRecord(record, userSpecifiedTypes, domainForExternalTypes);
+            default:
+                return new Pair<>(READ_ERROR_UNSUPPORTED_KEY, false);
+        }
+    }
+
+    private static Pair<String,Boolean> handleWellKnownTypeRecord(NdefRecord record,
+                                                                  String[] userSpecifiedTypes) {
+        if (Arrays.equals(record.getType(), NdefRecord.RTD_TEXT)) {
+            if (Arrays.asList(userSpecifiedTypes).contains(("text"))) {
+                return new Pair<>(readValueFromTextRecord(record), true);
+            } else {
+                return new Pair<>(READ_ERROR_MISMATCH_KEY, false);
+            }
+        } else {
+            return new Pair<>(READ_ERROR_UNSUPPORTED_KEY, false);
+        }
+    }
+
+    private static Pair<String,Boolean> handleExternalTypeRecord(NdefRecord record,
+                                                                 String[] userSpecifiedTypes,
+                                                                 String domainForExternalTypes) {
+        String typeAsString = new String(record.getType(), UTF8_CHARSET);
+        if (recordTypeIsExpected(typeAsString, userSpecifiedTypes, domainForExternalTypes)) {
+            return new Pair<>(new String(record.getPayload(), UTF8_CHARSET), true);
+        } else {
+            return new Pair<>(READ_ERROR_MISMATCH_KEY, false);
+        }
+    }
+
+    private static boolean recordTypeIsExpected(String recordTypeAsString,
+                                                String[] userSpecifiedTypes,
+                                                String domainForExternalTypes) {
+        for (String type : userSpecifiedTypes) {
+            type = domainForExternalTypes + ":" + type;
+            if (type.equals(recordTypeAsString)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static String readValueFromTextRecord(NdefRecord textTypeRecord) {
+        byte[] fullPayload = textTypeRecord.getPayload();
+        // The payload includes a prefix denoting the language, so we need to parse that off
+        int langBytesLength = fullPayload[0]; // status byte
+        int lengthOfPrefix = langBytesLength + 1; // add 1 for the status byte itself
+        byte[] payloadWithoutLang = Arrays.copyOfRange(fullPayload, lengthOfPrefix, fullPayload.length);
+        return new String(payloadWithoutLang, UTF8_CHARSET);
+    }
+
+    protected static NdefRecord createNdefRecord(String userSpecifiedType,
+                                                 String userSpecifiedDomain,
+                                                 String payloadToWrite) {
+        if (isCommCareSupportedWellKnownType(userSpecifiedType)) {
+            return createWellKnownTypeRecord(userSpecifiedType, payloadToWrite);
```

Comment:
Can you explain (maybe comment) the effective difference between a "well known" and "external" type record?

Reply:
The confluence documentation that I linked to explains this. Specifically the "Choosing and Defining Payload Types" section, although reading the whole page would probably be useful context.